### PR TITLE
Add cucim.kit.cumed plugin with skeleton

### PR DESCRIPTION
**Pull request by gigony**
This change adds a skeleton plugin named `cucim.kit.cumed` for medical-related file loaders and image operations (e.g., .mhd file format support).

The image loader to use is determined by `is_valid()` method implemented in each plugin and CuImage object holds the reference to the image format descriptor (`cucim::io::format::ImageFormatDesc*`).

cuCIM treats `cucim.kit.cuslide` and `cucim.kit.cumed` plugins as built-in plugins, and it would support custom plugins in the future (by refactoring the framework code, according to the recent update of Carbonite (minimal) SDK library and use the library as it is).

### CODE REVIEW

File: cpp/include/cucim/core/framework.h [hunk 1]
```diff
@@ -63,7 +63,7 @@ struct PluginLoadingDesc
 
 struct Framework
 {
-    void load_plugins(const PluginLoadingDesc& desc = PluginLoadingDesc::get_default());
+    // void load_plugins(const PluginLoadingDesc& desc = PluginLoadingDesc::get_default());
```
Comment (jakirkham) [inline cpp/include/cucim/core/framework.h:66]: Should we drop this? Or if we want to add this back later, should we add an explanatory comment/issue to track?

File: cpp/include/cucim/plugin/image_format.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+namespace cucim::plugin
+{
+
+class ImageFormat
+{
+public:
+    ImageFormat() = default;
+    ~ImageFormat() = default;
+
+    bool add_interfaces(cucim::io::format::IImageFormat* image_formats);
```
Comment (jakirkham) [inline cpp/include/cucim/plugin/image_format.h:33]: Is it possible to use a reference here? Could the variable be `const`?

File: cpp/plugins/cucim.kit.cumed/src/cumed/cumed.cpp [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+    }
+
+    const size_t associated_image_count = 0;
+    std::pmr::vector<std::string_view> associated_image_names(&resource);
+
+    std::string_view raw_data{ "" };
+
+    // Dynamically allocate memory for json_data (need to be freed manually);
+    const std::string& json_str = std::string{};
+    char* json_data_ptr = static_cast<char*>(cucim_malloc(json_str.size() + 1));
```
Comment (jakirkham) [inline cpp/plugins/cucim.kit.cumed/src/cumed/cumed.cpp:187]: Is there a corresponding `free` somewhere? Would it be possible to use RAII to handle this allocation?

File: cpp/plugins/cucim.kit.cumed/tests/test_basic.cpp [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+#include "config.h"
+
+#include <catch2/catch.hpp>
+#include <chrono>
+
+TEST_CASE("Verify file", "[test_basic.cpp]")
+{
+    REQUIRE(1 == 1);
```
Comment (jakirkham) [inline cpp/plugins/cucim.kit.cumed/tests/test_basic.cpp:24]: Were we wanting to add more here?

**Review by jakirkham — COMMENTED** [review]
Thanks Gigon! 😄

Had a couple questions below

File: cpp/include/cucim/core/framework.h (hunk 1 above)
Comment (gigony) [reply cpp/include/cucim/core/framework.h:66]: Updated a comment and created an issue for handling that.
- https://github.com/rapidsai/cucim/issues/134

File: cpp/include/cucim/plugin/image_format.h (hunk 2 above)
Comment (gigony) [reply cpp/include/cucim/plugin/image_format.h:33]: Thanks for pointing it out!
Updated to use `const` variable.

File: cpp/plugins/cucim.kit.cumed/src/cumed/cumed.cpp (hunk 3 above)
Comment (gigony) [reply cpp/plugins/cucim.kit.cumed/src/cumed/cumed.cpp:187]: Yes, it is at the destructor of CuImage class.

https://github.com/rapidsai/cucim/blob/95e44833752209b548c1b5e7aa40217d0187909b/cpp/src/cuimage.cpp#L236-L241

I have created an issue to handle general resource handling issues and will follow up on the improvement through the issue.
- https://github.com/rapidsai/cucim/issues/135

File: cpp/plugins/cucim.kit.cumed/tests/test_basic.cpp (hunk 4 above)
Comment (gigony) [reply cpp/plugins/cucim.kit.cumed/tests/test_basic.cpp:24]: Currently, it is just a placeholder to implement unit tests for cumed plugin.

**Comment by gigony** [comment]
@ajschmidt8 It seems that it has an issue with cuda11.5 (same for https://github.com/rapidsai/cucim/pull/128)

https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci-cuda115/job/cucim/job/prb/job/cucim-gpu-test/CUDA=11.5,GPU_LABEL=cuda115,LINUX_VER=centos7,PYTHON=3.8/3/console

```bash
19:01:04 ++ python -c 'import cucim'
19:01:05 ci/gpu/build.sh: line 91:  1680 Segmentation fault      (core dumped) python -c 'import cucim'
19:01:05 Build step 'Execute shell' marked build as failure
19:01:08 Recording test results
```
Is there a known issue regarding this?

**Comment by ajschmidt8** [comment]
> @ajschmidt8 It seems that it has an issue with cuda11.5 (same for #128)
> 
> https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci-cuda115/job/cucim/job/prb/job/cucim-gpu-test/CUDA=11.5,GPU_LABEL=cuda115,LINUX_VER=centos7,PYTHON=3.8/3/console
> 
> ```shell
> 19:01:04 ++ python -c 'import cucim'
> 19:01:05 ci/gpu/build.sh: line 91:  1680 Segmentation fault      (core dumped) python -c 'import cucim'
> 19:01:05 Build step 'Execute shell' marked build as failure
> 19:01:08 Recording test results
> ```
> 
> Is there a known issue regarding this?
> 
> Update: It seems so. I can see failures in other projects - [rapidsai/cudf#9562](https://github.com/rapidsai/cudf/pull/9562)

Yes, `11.5` tests are optional right now and won't prevent merging. We're still working out some build issues. Any errors unrelated to this repo are probably fine to ignore for now.

**Comment by gigony** [comment]
Thanks @ajschmidt8 for the confirmation!

**Review by jakirkham — APPROVED** [review]

**Pull request merged**